spec/script_spec.rb

```diff
@@ -0,0 +1,237 @@
+require_relative '../script.rb'
+
+describe Enumerable do
+  let(:arraio) { [1, 2, 3, 4, 5] }
+  let(:rangio) { (1..5) }
+  let(:blocko) { |element| element }
+  let(:hasho) { {} }
+  let(:stringos) { %w[cat dog wombat] }
+  let(:block1) { proc { |x| x.length >= 3 } }
+  let(:block2) { proc { |x| x.length >= 4 } }
+  let(:words) { %w[ant bear cat] }
+  let(:hashproc) { proc { |item, index| hasho[item] = index } }
+
+  describe '#my_each' do
+    context 'when no block is passed' do
+      it 'returns an enumerator range when called against a range' do
+        expect(rangio.my_each.class).to eql(arraio.each.class)
+      end
+
+      it 'returns an Enumerator array when called against an array' do
+        expect(arraio.my_each.class).to eql(arraio.each.class)
+      end
+    end
+
+    context 'when a block is passed' do
+      it 'calls the given block once for each element in self and returns the array itself' do
+        expect(arraio.my_each { blocko }).to eql(arraio.each { blocko })
+      end
+
+      it 'calls the given block once for each element in self and returns the range itself' do
+        expect(rangio.my_each { blocko }).to eql(rangio.each { blocko })
+      end
+    end
+  end
+
+  describe '#my_each_with_index' do
+    context 'when no block is passed' do
+      it 'returns an enumerator' do
+        expect(arraio.my_each_with_index.class).to eql(arraio.each_with_index.class)
+      end
+    end
+
+    context 'when a block is passed' do
+      it 'calls the item and its index for each item in enum' do
+        expect(stringos.my_each_with_index(&hashproc)).to eql(stringos.each_with_index(&hashproc))
+      end
+    end
+  end
+
+  describe '#my_select' do
+    let(:proco) { proc { |num| num.even? } }
+
+    context 'when no block is given' do
+      it 'returns an Enumerator' do
+        expect(arraio.my_select.class).to eql(arraio.select.class)
+      end
+    end
+
+    context 'when a block is given' do
+      it 'returns an array for all element for which the block returns a true value' do
+        expect(arraio.my_select(&proco)).to eql([2, 4])
+      end
+    end
+  end
+
+  describe '#my_all?' do
+    let(:number) { [1, 2i, 3.14] }
+
+    context 'when block is given' do
+      it 'returns true if the block never returns false or nil' do
+        expect(words.my_all?(&block1)).to eql(true)
+      end
+      it 'returns true if the block never returns false or nil' do
+        expect(words.my_all?(&block2)).to eql(false)
+      end
```

Comment:
Here you mentioned that `returns true if the block never returns false or nil` But you are expecting to return false in this scenario. Don't you think that the message should have a co-relation between expectation and the statement? I would rather say `returns false if the block returns false or nil`.

Reply:
Yes, that makes sense. Thanks!